libc/stdlib_alloc.c

```diff
@@ -0,0 +1,561 @@
+#else
+	TEST_ASSERT_NOT_NULL(ptr);
+#endif
+
+	int *ptr1 = calloc(BLOCK_SIZE, 0);
+#ifdef __phoenix__
+	TEST_ASSERT_NULL(ptr1);
+#else
+	TEST_ASSERT_NOT_NULL(ptr1);
+#endif
+
+	int *ptr2 = calloc(0, 0);
+#ifdef __phoenix__
+	TEST_ASSERT_NULL(ptr2);
+#else
+	TEST_ASSERT_NOT_NULL(ptr2);
+#endif
+
+	free(ptr);
+	free(ptr1);
+	free(ptr2);
+}
+
+
+TEST(stdlib_malloc, calloc_one)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, *ptr);
+#ifndef __phoenix__
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+
+TEST(stdlib_malloc, calloc_large)
+{
+	int *ptr = calloc(BLOCK_SIZE * LARGE_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE * LARGE_SIZE);
+#ifndef __phoenix__
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+
+TEST(stdlib_malloc, calloc_iterate)
+{
+	for (size_t s = 1; s <= 4096; s++) {
+		int *ptr = calloc(1, s * 10);
+		TEST_ASSERT_NOT_NULL(ptr);
+		TEST_ASSERT_EQUAL_INT(0, *ptr);
+#ifndef __phoenix__
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(s * 10, malloc_usable_size(ptr));
+#endif
+		free(ptr);
+	}
+}
+
+
+TEST(stdlib_malloc, calloc_overflow)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Walloc-size-larger-than="
+
+	errno = 0;
+	TEST_ASSERT_NULL(calloc(1, SIZE_MAX));
+	TEST_ASSERT_EQUAL_INT(ENOMEM, errno);
+
+	errno = 0;
+	TEST_ASSERT_NULL(calloc(SIZE_MAX, 1));
+	TEST_ASSERT_EQUAL_INT(ENOMEM, errno);
+
+	errno = 0;
+	TEST_ASSERT_NULL(calloc(SIZE_MAX, SIZE_MAX));
+	TEST_ASSERT_EQUAL_INT(ENOMEM, errno);
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(stdlib_malloc, realloc_null)
+{
+	int *ptr = realloc(NULL, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	int *ptr1 = realloc(NULL, 0);
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	free(ptr);
+	free(ptr1);
+}
+
+
+TEST(stdlib_malloc, realloc_zero_size)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, *ptr);
+#ifndef __phoenix__
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr));
+#endif
+
+	int *ptr1 = realloc(ptr, 0);
+	TEST_ASSERT_NULL(ptr1);
+}
+
+
+TEST(stdlib_malloc, realloc_calloc_resize)
+{
+	int NUMBERS[BLOCK_SIZE];
+	int *ptr = calloc(BLOCK_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE);
+	for (int i = 0; i < BLOCK_SIZE; i++) {
+		ptr[i] = i;
+		NUMBERS[i] = i;
+	}
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr, BLOCK_SIZE);
+	TEST_ASSERT_EQUAL_INT(0, (uintptr_t)ptr % sizeof(int));
+
+	int *ptr1 = realloc(ptr, BLOCK_SIZE * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr1, BLOCK_SIZE);
+	TEST_ASSERT_EQUAL_INT(0, (uintptr_t)ptr1 % sizeof(int));
+
+	free(ptr1);
+}
+
+
+TEST(stdlib_malloc, realloc_calloc_resize_smaller)
+{
+	int *ptr = calloc(1, 2 * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, ptr[0]);
```

Comment:
Check both of ptr[0] and ptr[1] so I suggest using EACH_EQUAL and maybe change to `calloc(2, sizeof(int))` to emphasize that this is a memory block of two ints.